File: msd_radix_sort.go
package main

import "fmt"

const base = 10

func radixSort(array []int) []int {

	maxValue := max(array)
	// leftmost place value
	placeValue := pow(base, getNumberOfDigits(maxValue)-1)

	bucketSort(array, placeValue)

	return array
}

func bucketSort(array []int, placeValue int) []int {
	if placeValue < 1 {
		return array
	}

	if len(array) <= 1 {
		return array
	}

	buckets := [][]int{}
	minValue := getMinValue(array, placeValue)

	offset := 0
	if minValue < 0 {
		offset = 0 - minValue
	}
	for i := 0; i < (base + offset); i++ {
		buckets = append(buckets, []int{})
	}

	for idx, num := range array {
		bucketIdx := getDigit(array, idx, placeValue) + offset
		buckets[bucketIdx] = append(buckets[bucketIdx], num)
	}
	fmt.Println("place value: ", placeValue, "bucket: ", buckets)

	sortedIdx := 0
	for _, bucket := range buckets {
		sortedBucket := bucketSort(bucket, placeValue/base)
		for _, num := range sortedBucket {
			array[sortedIdx] = num
			sortedIdx += 1
		}
	}
	fmt.Println("place value: ", placeValue, "array: ", array)

	return array
}

func getNumberOfDigits(num int) int {
	numberOfDigits := 0
	for num > 0 {
		numberOfDigits += 1
		num /= base
	}
	return numberOfDigits
}

func abs(value int) int {
	if value < 0 {
		return -value
	}
	return value
}

func max(array []int) int {
	maxValue := array[0]
	for i := 1; i < len(array); i++ {
		if array[i] > maxValue {
			maxValue = array[i]
		}
	}
	return maxValue
}

func pow(a int, power int) int {
	result := 1
	for i := 0; i < power; i++ {
		result *= a
	}
	return result
}

// get minimum value of each digit group
func getMinValue(array []int, placeValue int) int {
	minValue := getDigit(array, 0, placeValue)
	for i := 1; i < len(array); i++ {
		element := getDigit(array, i, placeValue)
		if minValue > element {
			minValue = element
		}
	}
	return minValue
}

// ex. number: 2022, place value: 100 => digit = 0
// ex. number: -1223, place value: 100 => digit = -2
func getDigit(array []int, idx int, placeValue int) int {
	element := array[idx]
	digit := abs(element) / placeValue % base
	if element < 0 {
		return -digit
	}
	return digit
}

func main() {
	arrays := [][]int{
		{2022, 5, 398, 159, 16, 1223, 7},
		{2022, -5, 398, 159, 16, -1223, 7},
	}

	for _, array := range arrays {
		fmt.Println("unsorted array: ", array)
		fmt.Println("sorted array: ", radixSort(array))
		fmt.Println()
	}
}

/* outout
unsorted array:  [2022 5 398 159 16 1223 7]
place value:  1000 bucket:  [[5 398 159 16 7] [1223] [2022] [] [] [] [] [] [] []]
place value:  100 bucket:  [[5 16 7] [159] [] [398] [] [] [] [] [] []]
place value:  10 bucket:  [[5 7] [16] [] [] [] [] [] [] [] []]
place value:  1 bucket:  [[] [] [] [] [] [5] [] [7] [] []]
place value:  1 array:  [5 7]
place value:  10 array:  [5 7 16]
place value:  100 array:  [5 7 16 159 398]
place value:  1000 array:  [5 7 16 159 398 1223 2022]
sorted array:  [5 7 16 159 398 1223 2022]

unsorted array:  [2022 -5 398 159 16 -1223 7]
place value:  1000 bucket:  [[-1223] [-5 398 159 16 7] [] [2022] [] [] [] [] [] [] []]
place value:  100 bucket:  [[-5 16 7] [159] [] [398] [] [] [] [] [] []]
place value:  10 bucket:  [[-5 7] [16] [] [] [] [] [] [] [] []]
place value:  1 bucket:  [[-5] [] [] [] [] [] [] [] [] [] [] [] [7] [] []]
place value:  1 array:  [-5 7]
place value:  10 array:  [-5 7 16]
place value:  100 array:  [-5 7 16 159 398]
place value:  1000 array:  [-1223 -5 7 16 159 398 2022]
sorted array:  [-1223 -5 7 16 159 398 2022]
*/

File: msd_radix_sort.py
BASE = 10

def radixSort(array):
    if len(array) == 1:
        return array
    
    maxValue = max(array)
    # leftmost place value
    placeValue = BASE**(getNumberOfDigits(maxValue) - 1)

    bucketSort(array, placeValue)
       
    return array


def getNumberOfDigits(num):
    numberOfdigits = 0
    while num > 0:
        numberOfdigits += 1
        num /= BASE
    return numberOfdigits

# recursion
def bucketSort(array, placeValue):
    if placeValue < 1:
        return array
    
    if len(array) <= 1:
        return array

    buckets = []
    minValue = getMinValue(array, placeValue)

    offset = 0
    if minValue < 0:
        offset = 0 - minValue
    for i in range(BASE+offset):
        buckets.append([])

    for idx, num in enumerate(array):
        bucketIdx = getDigit(array, idx, placeValue) + offset
        buckets[bucketIdx].append(num)
    print("place value: ", placeValue, "bucket: ",  buckets)
    

    sortedIdx = 0
    for bucket in buckets:
        sortedBucket = bucketSort(bucket, placeValue/BASE)
        for num in sortedBucket:
            array[sortedIdx] = num
            sortedIdx += 1
    print("place value: ", placeValue, "array: ",  array)

    return array

def getMinValue(array, placeValue):
    minValue = getDigit(array, 0, placeValue)
    for i in range(1, len(array)):
        element = getDigit(array, i, placeValue)
        if element < minValue:
            minValue = element
    return minValue

def getDigit(array, idx, placeValue):
    element = array[idx]
    digit = abs(element) // placeValue % BASE
    return digit if element > 0 else -digit

if __name__ == "__main__":
    arrays = [[2022, 5, 398, 159, 16, 1223, 7], [2022, -5, 398, 159, 16, -1223, 7]]
    for array in arrays:
        print("unsorted array: ", array)
        print("sorted array: ", radixSort(array))
        print("\n")
    
"""
'unsorted array: ', [2022, 5, 398, 159, 16, 1223, 7])
('place value: ', 1000, 'bucket: ', [[5, 398, 159, 16, 7], [1223], [2022], [], [], [], [], [], [], []])
('place value: ', 100, 'bucket: ', [[5, 16, 7], [159], [], [398], [], [], [], [], [], []])
('place value: ', 10, 'bucket: ', [[5, 7], [16], [], [], [], [], [], [], [], []])
('place value: ', 1, 'bucket: ', [[], [], [], [], [], [5], [], [7], [], []])
('place value: ', 1, 'array: ', [5, 7])
('place value: ', 10, 'array: ', [5, 7, 16])
('place value: ', 100, 'array: ', [5, 7, 16, 159, 398])
('place value: ', 1000, 'array: ', [5, 7, 16, 159, 398, 1223, 2022])
('sorted array: ', [5, 7, 16, 159, 398, 1223, 2022])


('unsorted array: ', [2022, -5, 398, 159, 16, -1223, 7])
('place value: ', 1000, 'bucket: ', [[-1223], [-5, 398, 159, 16, 7], [], [2022], [], [], [], [], [], [], []])
('place value: ', 100, 'bucket: ', [[-5, 16, 7], [159], [], [398], [], [], [], [], [], []])
('place value: ', 10, 'bucket: ', [[-5, 7], [16], [], [], [], [], [], [], [], []])
('place value: ', 1, 'bucket: ', [[-5], [], [], [], [], [], [], [], [], [], [], [], [7], [], []])
('place value: ', 1, 'array: ', [-5, 7])
('place value: ', 10, 'array: ', [-5, 7, 16])
('place value: ', 100, 'array: ', [-5, 7, 16, 159, 398])
('place value: ', 1000, 'array: ', [-1223, -5, 7, 16, 159, 398, 2022])
('sorted array: ', [-1223, -5, 7, 16, 159, 398, 2022])
"""
